src/Effectful/Resource.hs

```diff
@@ -100,6 +106,22 @@ registerEff release = do
     es1 <- cloneEnv es0
     RI.register' istate $ unEff release es1
 
+-- | A variant of 'R.release' adjusted to work in the 'Eff' monad.
+releaseEff :: Resource :> es => R.ReleaseKey -> Eff es ()
+releaseEff = unsafeEff_ . R.release
+
+-- | A variant of 'R.unprotect' adjusted to work in the 'Eff' monad.
+--
+-- /Note:/ if the resource was acquired using 'allocateEff', 'allocateEff_' or
+-- 'registerEff' then the @release@ action returned will run in a clone of the
+-- environment it was registered in, so the effect row @es'@ will be ignored.
+-- Please see the documentation of the aforementioned functions.
+unprotectEff
```

Comment:
> One issue is that IIUC any invocation of this function without indicating what `es'` is will result in a compilation error about ambiguous type.

I am afraid I don't understand what you mean. The following compiles just fine:
```haskell
{-# LANGUAGE FlexibleContexts #-}
{-# LANGUAGE TypeOperators #-}

module Main (main) where

import Data.Maybe (fromMaybe)
import Effectful
import Effectful.Dispatch.Static (unsafeEff_)
import Effectful.Resource

main :: IO ()
main = runEff . runResource $ do
    releaseMe <- myUnprotected
    fromMaybe (pure ()) releaseMe

myUnprotected :: (Resource :> es, Resource :> es') => Eff es (Maybe (Eff es' ()))
myUnprotected = do
    (key, _) <- myAllocation
    unprotectEff key

myAllocation :: (Resource :> es) => Eff es (ReleaseKey, ())
myAllocation =
    allocateEff
        (unsafeEff_ $ putStrLn "Allocating")
        (\_ -> unsafeEff_ $ putStrLn "Releasing")
```

> 
> I think a signature `unprotectEff :: Resource :> es => R.ReleaseKey -> Eff es (Maybe (IO ()))` would be more reasonable here 🤔 Thoughts?

I tried to avoid the `IO` action because the consuming side has to use either `liftIO` which introduces an unnecessary `IOE :> es` constraint or `unsafeEff_` which is also not ideal.
With the `Resource :> es'` constraint it is clear from the type signature of the consuming function that some resource handling is involved.
One other alternative I came up with is returning a `Eff '[Resource] ()`. Then the consumer can use that with `inject`. It works, but it's not really elegant.

That was at least the idea behind that implementation; If you're still not convinced I am fine with changing that.

Reply:
> I am afraid I don't understand what you mean. The following compiles just fine:

Yeah, because you're immediately using the `releaseMe` in the same effect stack, so GHC unifies `es` with `es'` and it all works out.

Now, see what happens if you replace `fromMaybe (pure ()) releaseMe` with `pure ()`, i.e. leave `releaseMe` unused:

```
    • Ambiguous type variable ‘es'0’ arising from a use of ‘myUnprotected’
      prevents the constraint ‘(Resource :> es'0)’ from being solved.
      Probable fix: use a type annotation to specify what ‘es'0’ should be.
      Potentially matching instances:
        instance [overlapping] e :> (e : es)
          -- Defined in ‘Effectful.Internal.Effect’
        instance (e :> es) => e :> (x : es)
          -- Defined in ‘Effectful.Internal.Effect’
        ...plus one other
        (use -fprint-potential-instances to see them all)
    • In a stmt of a 'do' block: releaseMe <- myUnprotected
      In the second argument of ‘($)’, namely
        ‘do releaseMe <- myUnprotected
            pure ()’
      In the expression:
        runEff . runResource
          $ do releaseMe <- myUnprotected
               pure ()
    |
162 |   releaseMe <- myUnprotected
    |                ^^^^^^^^^^^^^
```

> One other alternative I came up with is returning a Eff '[Resource] (). Then the consumer can use that with inject. It works, but it's not really elegant.

Yeah, this is better, but not very elegant as you say.

I don't really know what the best solution is to this issue and I don't want to delay this PR any further, so let's maybe agree to leave out `unprotectEff` for now from it? Then I can merge it and make a release and we can tackle the problem of `unprotectEff` separately.

Reply:
Ah, ok, now I get what you meant! Yes, I agree that this might not lead to the best user experience.

> I don't really know what the best solution is to this issue and I don't want to delay this PR any further, so let's maybe agree to leave out `unprotectEff` for now from it? Then I can merge it and make a release and we can tackle the problem of `unprotectEff` separately.

Agreed. I'll move the `unprotectEff` to a separate PR :+1: